data/src/main/java/com/linkedin/data/schema/PathSpecSet.java

```diff
@@ -0,0 +1,377 @@
+   * {@code PathSpecSet.of(Zing.fields().foo().bar()}.
+   *
+   * However, suppose you already have a {@link PathSpecSet} from the perspective of "Foo" but need a
+   * {@link PathSpecSet} for your "Zing" downstream.  This method make this easy:
+   * {@code
+   *   PathSpecSet fooPathSpecSet = PathSpecSet.of(Foo.fields().bar());
+   *   PathSpecSet zingPathSpecSet = fooPathSpecSet.copyWithScope(Zing.fields().foo());
+   *
+   *   zingPathSpecSet.equals(PathSpecSet.of(Zing.fields().foo().bar()); // true
+   * }
+   *
+   * If you scope an empty {@link PathSpecSet} it remains empty.
+   *
+   * @param parent the parent {@link PathSpec} to use when scoping the contained {@link PathSpec}s
+   * @return a new {@link PathSpecSet} that is scoped to the new parent
+   */
+  public PathSpecSet copyWithScope(PathSpec parent) {
+    if (this.isFullProjection()) {
+      return PathSpecSet.of(parent);
+    }
+
+    if (this.isEmpty()) {
+      return PathSpecSet.empty();
+    }
+
+    Builder builder = newBuilder();
+
+    this.getPathSpecs().stream()
+        .map(childPathSpec -> {
+          ArrayList<String> list = new ArrayList<>();
+          list.addAll(parent.getPathComponents());
+          list.addAll(childPathSpec.getPathComponents());
+          return list;
+        })
+        .map(PathSpec::new)
+        .forEach(builder::add);
+
+    return builder.build();
+  }
+
+  /**
+   * Return a copy of this {@link PathSpecSet} where only {@link PathSpec}s that are prefixed with the input {@link PathSpec} are retained.
+   *
+   * Additionally, the prefix is removed for the retained {@link PathSpec}s.
+   *
+   * Here are some examples showing the functionality:
+   *
+   * {@code
+   *   // This PathSpecSet is empty because no PathSpecs originally contained start with "abc"
+   *   PathSpecSet emptyPathSpecSet = PathSpecSet.of(Foo.fields().bar().baz()).copyAndRemovePrefix(Foo.fields().abc());
+   *
+   *   // This PathSpecSet is full because it contains the entire prefix PathSpec
+   *   PathSpecSet fullPathSpecSet = PathSpecSet.of(Foo.fields().bar()).copyAndRemovePrefix(Foo.fields().bar())
+   *
+   *   // The following "equals" evaluates to true
+   *   PathSpecSet prefixRemovedPathSpecSet = PathSpecSet.of(Foo.fields().bar().baz(), Foo.fields().bar().abc()).copyAndRemovePrefix(Foo.fields().bar());
+   *   prefixRemovedPathSpecSet.equals(PathSpecSet.of(new PathSpec("baz"), new PathSpec("abc")));
+   * }
+   *
+   * @param prefix the {@link PathSpec} prefix to use when retaining {@link PathSpec}s.
+   * @return a {@link PathSpecSet} with elements starting with the input {@link PathSpec} prefix
+   */
+  public PathSpecSet copyAndRemovePrefix(PathSpec prefix) {
+    if (isFullProjection() || isEmpty()) {
+      // full or empty projections stay the same
+      return this;
+    }
+
+    // if we contain the exact prefix PathSpec, it should be a full projection
+    if (getPathSpecs().contains(prefix)) {
```

Comment:
This should check for the prefix or any parent of the perfix, right?

Reply:
yes you are correct. I'll modify accordingly